### PR TITLE
feat(clique): dynamically adjust block time on single-signer network through geth console

### PR DESCRIPTION
**Pull request by NazariiDenha**
## 1. Purpose or design rationale of this PR

Description of task: https://www.notion.so/scrollzkp/l2geth-dynamic-block-time-3116d757b2a343f7a7dc9feed5273284
Design: https://www.notion.so/scrollzkp/l2geth-dynamic-block-time-9a46d99bfb744210b65fad3956a98e46?pvs=4

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [x] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [ ] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [x] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

File: consensus/clique/clique.go [hunk 1]
```diff
@@ -191,16 +191,15 @@ type Clique struct {
 // signers set to the ones provided by the user.
 func New(config *params.CliqueConfig, db ethdb.Database) *Clique {
 	// Set any missing consensus parameters to their defaults
-	conf := *config
```
Comment (Thegaram) [inline consensus/clique/clique.go:194]: Any reason they added this `conf := *config` line before?

File: internal/web3ext/web3ext.go [hunk 2]
```diff
@@ -68,6 +68,16 @@ web3._extend({
 			call: 'clique_discard',
 			params: 1
 		}),
+		new web3._extend.Method({
+			name: 'getPeriod',
+			call: 'clique_getPeriod',
+			params: 0
+		}),
```
Comment (Thegaram) [inline internal/web3ext/web3ext.go]: You can consider adding this under `properties`.

File: consensus/clique/clique.go (hunk 1 above)
Comment (NazariiDenha) [reply consensus/clique/clique.go:194]: Don't know any suitable reason. `config` is not changed in any places inside `clique` struct

File: internal/web3ext/web3ext.go (hunk 2 above)
Comment (NazariiDenha) [reply internal/web3ext/web3ext.go]: Didn't find how to get from nested structure

**Review by 0xmountaintop — DISMISSED** [review]

**Review by Thegaram — DISMISSED** [review]

**Comment by Thegaram** [comment]
Note: This API is not suitable for multi-signer networks. If `nodeA` increases the block time, it will start rejecting blocks from `nodeB`.

https://github.com/scroll-tech/go-ethereum/blob/84b567195aa5b950fb7ec19c2543d9a39690de02/consensus/clique/clique.go#L330-L332

Possible solutions:
- Be more lenient with the timestamp checking, e.g. accept blocks with the old clique period for a certain time period.
- Propose/accept the new block time through Clique consensus (similar to how we can add/remove signers).

**Review by Thegaram — COMMENTED** [review]
Note: This API is not suitable for multi-signer networks. If `nodeA` increases the block time, it will start rejecting blocks from `nodeB`. Even with just one signer, the follower nodes might fork if the signer decreases the block period.

https://github.com/scroll-tech/go-ethereum/blob/84b567195aa5b950fb7ec19c2543d9a39690de02/consensus/clique/clique.go#L330-L332

The same applies to syncing the chain from genesis. If we changed the block time several times in the past, the syncing node would need to use the shortest block time, or it would reject blocks.

Possible solutions:
- Be more lenient with the timestamp checking, e.g. accept blocks with the old clique period for a certain time period.
- Propose/accept the new block time through Clique consensus (similar to how we can add/remove signers).

Another thing that would become an issue: After restarting the node, we'll use the old clique period (read from DB), the updates through the console are not persisted.

---

We need to think more about these issues before we merge this feature.

**Comment by NazariiDenha** [comment]
I will convert it to a draft to prevent an accidental merge

**Pull request closed without merging**